### PR TITLE
refactor: Primer -> Oligo, PrimerLike -> OligoLike

### PR DESCRIPTION
**Pull request by emmcauley**
The PR refactors the `Primer` object and planned `Probe` object into the same dataclass, now called `Oligo`. An `Oligo` object can represent single primer and probe designs. The corresponding base class, previously called `PrimerLike`, has been refactored to `OligoLike`. 

The new `Oligo` object has the attributes that a `Primer` object had (`tm`, `penalty`, `span` and optional attributes `bases`, `tail`) _plus_ three new optional attributes: `self_any_th`, `self_end_th`, and `hairpin_th`. These three attributes are emitted from Primer3 for probe design and, after an update to `PrimerAndAmpliconParams`, will also be emitted for primer designs.   

For backwards compatibility, the `Primer` object now wraps an `Oligo` object and also emits a `DeprecationWarning`.

I left `PrimerPair` largely untouched -- a `PrimerPair` is now composed of two `Oligo` objects but the rest of the functionality of `PrimerPair` I left intact.

### CODE REVIEW

File: prymer/primer3/primer3.py [hunk 1]
```diff
@@ -371,12 +365,18 @@ def design_primers(self, design_input: Primer3Input) -> Primer3Result:  # noqa:
...
-
-        design_region: Span = self._create_design_region(
-            target_region=design_input.target,
-            max_amplicon_length=design_input.params.max_amplicon_length,
-            min_primer_length=design_input.params.min_primer_length,
-        )
+        design_region: Span
+        match design_input.task:
+            case PickHybProbeOnly():
+                design_region = design_input.target
```
Comment (tfenne) [inline prymer/primer3/primer3.py:371]: We should likely validate that `target` is at least as long as the min probe length.

File: prymer/primer3/primer3.py [hunk 2]
```diff
@@ -534,51 +525,44 @@ def _build_primers(
...
-            unfiltered_designs: list of primers emitted from Primer3
-             design_results: key-value pairs of results reported by Primer3
+        """Screens oligo designs (primers or probes) emitted by Primer3 for acceptable dinucleotide
+        runs and extracts failure reasons for failed designs."""
 
-        Returns:
-            primer_designs: a `Primer3Result` that encapsulates valid and failed designs
-        """
-        valid_primer_designs = [
+        valid_oligo_designs = [
```
Comment (tfenne) [inline prymer/primer3/primer3.py:549]: or just `valid_designs`?

File: prymer/primer3/primer3.py [hunk 3]
```diff
@@ -534,51 +525,44 @@ def _build_primers(
...
-            if not Primer3._is_valid_primer(primer_design=design, design_input=design_input)
+            if not _has_acceptable_dinuc_run(oligo_design=design, design_input=design_input)
         ]
 
         failure_strings = [design_results[f"PRIMER_{design_input.task.task_type}_EXPLAIN"]]
         failures = Primer3._build_failures(dinuc_failures, failure_strings)
-        primer_designs: Primer3Result = Primer3Result(
-            filtered_designs=valid_primer_designs, failures=failures
+        design_candidates: Primer3Result = Primer3Result(
+            filtered_designs=valid_oligo_designs, failures=failures
```
Comment (tfenne) [inline prymer/primer3/primer3.py:563]: This came up elsewhere in conversation that perhaps `filtered_designs` should just be `designs`?  If `primer3` implemented a dinuc check, would we feel the need to call this `filtered_designs`?

File: prymer/primer3/primer3_input.py [hunk 4]
```diff
@@ -77,26 +90,52 @@
...
+    probe design."""
 
     target: Span
     task: Primer3TaskType
-    params: Primer3Parameters
-    weights: Primer3Weights = Primer3Weights()
+    primer_and_amplicon_params: Optional[PrimerAndAmpliconParameters] = None
+    probe_params: Optional[ProbeParameters] = None
+    primer_weights: Optional[PrimerAndAmpliconWeights] = None
+    probe_weights: Optional[ProbeWeights] = None
```
Comment (tfenne) [inline prymer/primer3/primer3_input.py:123]: Is there a reason not to allow default weights?  I.e. does primer3 blow up if you set probe weights and then design left primers?

File: prymer/primer3/primer3_parameters.py [hunk 5]
```diff
@@ -103,10 +154,15 @@ def __post_init__(self) -> None:
         if self.gc_clamp[0] > self.gc_clamp[1]:
             raise ValueError("Min primer GC-clamp must be <= max primer GC-clamp")
 
+        # Set melting temperature thresholds to be 10 degrees less than the minimum primer tm
+        default_thermo_tm: float = self.primer_tms.min - 10.0
+        object.__setattr__(self, "primer_max_self_any_thermo", default_thermo_tm)
+        object.__setattr__(self, "primer_max_self_end_thermo", default_thermo_tm)
+        object.__setattr__(self, "primer_max_hairpin_thermo", default_thermo_tm)
```
Comment (tfenne) [inline prymer/primer3/primer3_parameters.py:161]: Why not just do this above?

File: prymer/primer3/primer3_parameters.py [hunk 6]
```diff
@@ -92,6 +138,11 @@ class Primer3Parameters:
     primer_max_dinuc_bases: int = 6
     avoid_masked_bases: bool = True
     number_primers_return: int = 5
+    primer_max_self_any: float = 8.00
+    primer_max_self_any_thermo: float = field(init=False)
+    primer_max_self_end: float = 3.0
+    primer_max_self_end_thermo: float = field(init=False)
+    primer_max_hairpin_thermo: float = field(init=False)
```
Comment (tfenne) [inline prymer/primer3/primer3_parameters.py:145]: This feels like a rather unrelated addition - is there a reason this is part of this PR?

File: prymer/primer3/primer3_parameters.py [hunk 7]
```diff
@@ -145,3 +208,131 @@ def max_primer_length(self) -> int:
...
+        probe_max_self_any: the maximal local alignment score of aligning the probe to itself
+        probe_max_self_any_thermo: the maximal melting temperature of the most stable structure
+            resulting from aligning the probe to itself
+        probe_max_self_end: max allowable 3'-anchored global alignment score when testing a
+            probe for self-complementarity
+        probe_max_self_end_thermo: the maximal melting temperature of the most stable structure
+            resulting from aligning the 3' end of the probe
+        probe_max_hairpin_thermo: most stable monomer structure as calculated by a thermodynamic
+            approach
+        probe_excluded_region: the excluded region (start, length) that probes shall not overlap
```
Comment (tfenne) [inline prymer/primer3/primer3_parameters.py:246]: This does not feel like it belongs here - or if it does, it needs a lot more documentation on how to specify this in a target agnostic way.

**Review by tfenne — CHANGES_REQUESTED** [review]
As we're doing this we should also think about the future where we want to enable running primer3 in the mode where it will produce left and right primers _and_ a probe all in one go.

In the examples, can you please use more realistic values.  My 2c is that documentation that uses unrealistic values reduces my confidence in the documentation.  Examples are to use more realistic Tms, penalty values that are >=0, sequences that are not G*20 etc.

I don't love that this PR also introduces a ton of new parameters that we weren't previously exposing for the homodimer, end stability and hairpin thresholds ... I would ideally prefer to see those in a separate PR where we can discuss them a bit more.

File: prymer/primer3/primer3_input.py (hunk 4 above)
Comment (emmcauley) [reply prymer/primer3/primer3_input.py:123]: Yes.

File: prymer/primer3/primer3_parameters.py (hunk 5 above)
Comment (emmcauley) [reply prymer/primer3/primer3_parameters.py:161]: At this point in the code, you have validated other user inputs and you know that you have valid `PrimerAndAmpliconParameters`. It made more sense to me to set these attributes after you know that you have valid input and can continue.

File: prymer/primer3/__init__.py [hunk 8]
```diff
@@ -21,5 +22,6 @@
     "DesignPrimerPairsTask",
     "DesignRightPrimersTask",
     "Primer3Parameters",
```
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/primer3/__init__.py:24]: I removed `Primer3Parameters` and swapped in `PrimerAndAmpliconParameters` and `ProbeParameters` instead -- thanks for catching this!

File: prymer/primer3/primer3.py (hunk 3 above)
Comment (emmcauley) [reply prymer/primer3/primer3.py:563]: I have updated the docstring and changed the variable from `filtered_designs` to `designs`. These are filtered for out-of-spec characteristics and the corresponding `failures` list contains the reasons why (so I have been thinking of these designs as "filtered" because they have undergone one round of filtering for primer-specific properties like GC content, TM, etc.).

File: prymer/api/oligo.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+Optional attributes include naming information and a tail sequence to attach to the 5' end of the
+oligo (if applicable). Optional attributes also include the thermodynamic results from Primer3.
+
+## Examples of interacting with the `Oligo` class
+
+```python
+>>> from prymer.api.span import Span, Strand
+>>> oligo_span = Span(refname="chr1", start=1, end=20)
+>>> oligo = Oligo(tm=70.0, penalty=-123.0, span=oligo_span)
+>>> oligo.longest_hp_length()
```
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py:20]: I'm not sure I understand -- if `bases` is None, `longest_hp_length()` will return 0.

File: prymer/primer3/primer3.py (hunk 1 above)
Comment (emmcauley) [reply prymer/primer3/primer3.py:378]: this check is now added in `primer3.py` (it requires having access to the `ProbeParameters`, which we do not have in `primer3_task.py`)

File: prymer/primer3/primer3_parameters.py (hunk 7 above)
Comment (emmcauley) [reply prymer/primer3/primer3_parameters.py]: `probe_excluded_region` has been stripped out in the latest commit!

File: docs/overview.md [hunk 10]
```diff
@@ -18,22 +18,24 @@ amplicons prior to primer design.
 Designing primers (left or right) or primer pairs using Primer3 is primarily performed using the 
 [`Primer3`][prymer.primer3.primer3.Primer3] class, which wraps the
 [`primer3` command line tool](https://github.com/primer3-org/primer3).  The 
-[`design_primers()`][prymer.primer3.primer3.Primer3.design_primers] facilitates the design of single and paired primers 
+[`design_oligos()`][prymer.primer3.primer3.Primer3.design_oligos] facilitates the design of primers (single and paired) and internal hybridization probes
```
[the remark this answers is not in the document]
Comment (ameynert) [reply docs/overview.md:21]: Could it just be `design`?

File: prymer/api/oligo.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        return replace(self, name=name)
+
+    def bases_with_tail(self) -> Optional[str]:
+        """
+        Returns the sequence of the oligo prepended by the tail.
+
+        If either `bases` or `tail` are None, they shall be excluded. Return None if both are None.
+        """
+        if self.bases is None:
+            return None if self.tail is None else self.tail
```
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py]: as discussed via Slack, I think we'll handle this in another PR -- I'll leave this unresolved but wanted to flag this with my interpretation of our discussion

File: prymer/api/oligo.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    bases: Optional[str] = None
+    tail: Optional[str] = None
+
+    def __post_init__(self) -> None:
+        super(Oligo, self).__post_init__()
+
+    def longest_hp_length(self) -> int:
+        """Length of longest homopolymer in the oligo."""
+        if self.bases is None:
+            return 0
```
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py:128]: as discussed via Slack, I think we'll handle this in another PR -- I'll leave this unresolved but wanted to flag this with my interpretation of our discussion

File: prymer/api/oligo.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py:20]: as discussed via Slack, I think we'll handle this in another PR -- I'll leave this unresolved but wanted to flag this with my interpretation of our discussion

File: prymer/api/oligo.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
+"""
+# Oligo Class and Methods
+
+This module contains a class and class methods to represent an oligo designed by Primer3.
```
[the remark this answers is not in the document]
Comment (ameynert) [reply prymer/api/oligo.py:4]: Optional penalty would be nice for me, but we could do that in a subsequent PR.

File: docs/overview.md [hunk 14]
```diff
@@ -3,8 +3,8 @@
 The `prymer` Python library is intended to be used for three main purposes:
 
 1. [Clustering targets](#clustering-targets) into larger amplicons prior to designing primers.
-2. [Designing primers](#designing-primers) (left or right) or primer pairs using Primer3 for each target from (1).
-3. [Build and Picking a set of primer pairs](#build-and-picking-primer-pairs) from the designed primer pairs produced in (2).
+2. [Designing oligos](#designing-primers) using Primer3 for each target from (1).
```
[the remark this answers is not in the document]
Comment (msto) [reply docs/overview.md:6]: @emmcauley Let's revert this line and address more substantial documentation updates in subsequent PRs.

File: prymer/api/oligo.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (msto) [reply prymer/api/oligo.py:4]: I think it is a noble but perhaps lofty goal to expand the scope of this package beyond `primer3`, certainly beyond the scope of this PR.

I agree we should be mindful of making decisions that are overly specific to `primer3`. However I'd like to be cognizant that the content of this file is largely the same as the original `primer.py`, and would like to limit this PR's scope to the planned refactoring (`Primer` -> `Oligo`)

File: prymer/api/oligo.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (msto) [reply prymer/api/oligo.py:20]: Seconding Erin, and echoing my comment above - if the migration from `primer.py` to `oligo.py` calls attention to concerns in existing code, I think it's appropriate to identify them, but would like their resolution to be deferred to later PRs.

File: prymer/api/oligo.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (msto) [reply prymer/api/oligo.py:128]: Quoting my comment from above:

> if the migration from primer.py to oligo.py calls attention to concerns in existing code, I think it's appropriate to identify them, but would like their resolution to be deferred to later PRs.

File: prymer/api/oligo.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+    """
+
+    tm: float
+    penalty: float
+    span: Span
+    self_any: Optional[float] = None
+    self_any_th: Optional[float] = None
+    self_end: Optional[float] = None
+    self_end_th: Optional[float] = None
+    hairpin_th: Optional[float] = None
```
[the remark this answers is not in the document]
Comment (msto) [reply prymer/api/oligo.py:118]: @emmcauley I think this is the only functional change introduced in this PR. [^1]

Do you think you could remove the new attributes (so the PR is **only** renaming things from `Primer` to `Oligo`) and introduce a new PR with functional updates to accommodate probe design?

[^1]: If I'm wrong about this, please do the same for any other new code 🙂

File: prymer/primer3/primer3.py [hunk 16]
```diff
@@ -371,12 +365,25 @@ def design_primers(self, design_input: Primer3Input) -> Primer3Result:  # noqa:
...
+        match design_input.task:
+            case PickHybProbeOnly():
+                if design_input.target.length < design_input.probe_params.probe_sizes.min:
+                    raise ValueError(
+                        "Target region required to be at least as large as the"
+                        " minimal probe size: "
+                        f"target length: {design_input.target.length}, "
+                        f"minimal probe size: {design_input.probe_params.probe_sizes.min}"
+                    )
+                design_region = design_input.target
```
Comment (msto) [inline prymer/primer3/primer3.py:378]: @emmcauley This line caught my eye due to the lack of test coverage.

This looks like a functional change - I'd suggest backing this out to be introduced by another PR (e.g. #50 where `PickHybProbeOnly` is introduced).

File: prymer/primer3/primer3.py [hunk 17]
```diff
@@ -760,3 +747,44 @@ def _create_design_region(
...
+    Returns:
+
+    """
+    max_dinuc_bases: int
+    if design_input.task.requires_primer_amplicon_params:
+        max_dinuc_bases = design_input.primer_and_amplicon_params.primer_max_dinuc_bases
+    elif design_input.task.requires_probe_params:
+        max_dinuc_bases = design_input.probe_params.probe_max_dinuc_bases
+
+    return oligo_design.longest_dinucleotide_run_length() <= max_dinuc_bases
```
Comment (msto) [inline prymer/primer3/primer3.py:790]: Are these new?

**Review by msto — COMMENTED** [review]
Hey @tfenne @emmcauley 

I'm optimistic we can get this across the line in the next day or so.

@emmcauley I agree with Tim that this should be split up into a few PRs - one to execute the grand `Primer` -> `Oligo` renaming, and additional PR(s) to introduce the functional changes (e.g. `ProbeParameters`, the various new parameters and attributes)

@tfenne If this becomes a purely refactoring PR, I think we can merge in without much additional review. I agree the points you've highlighted are worth addressing, but I think we'll be able to progress more quickly if we keep the scope of this PR focused on the renaming.

File: docs/overview.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (tfenne) [reply docs/overview.md:21]: I'm fine with `design()`.

File: prymer/primer3/primer3_input.py (hunk 4 above)
Comment (tfenne) [reply prymer/primer3/primer3_input.py:123]: I find your answer confusing.  I'm asking "if you provide too many weights, isn't that ok?", but the last line of your answer - "it doesn't know how far off optimal it is when optimal is left undefined." - makes it sound like you're answering "what if we didn't provide enough weights"?

File: prymer/api/oligo.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+
+Optional attributes include naming information and a tail sequence to attach to the 5' end of the
+oligo (if applicable). Optional attributes also include the thermodynamic results from Primer3.
+
+## Examples of interacting with the `Oligo` class
+
+```python
+>>> from prymer.api.span import Span, Strand
+>>> oligo_span = Span(refname="chr1", start=1, end=20)
+>>> oligo = Oligo(tm=70.0, penalty=-123.0, span=oligo_span)
```
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py]: as discussed, I've opened issue #56 to be addressed in a future PR

File: prymer/api/oligo.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py:20]: as discussed, I've opened issue #55 to be addressed in a future PR (resolution of which should also simplify  `longest_hp_length()`)

File: prymer/api/oligo.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/oligo.py:128]: I've opened issue #55 (related)

File: prymer/api/primer_pair.py [hunk 19]
```diff
@@ -215,7 +215,7 @@ def to_bed12_row(self) -> str:
             )
         )
 
-    def __iter__(self) -> Iterator[Primer]:
+    def __iter__(self) -> Iterator[Oligo]:
```
[the remark this answers is not in the document]
Comment (emmcauley) [reply prymer/api/primer_pair.py:218]: as discussed, I've opened issue #58 to be addressed in a future PR

File: prymer/primer3/primer3_input.py (hunk 4 above)
Comment (emmcauley) [reply prymer/primer3/primer3_input.py:123]: To use your example: Primer3 blows up if you set probe weights and then design left primers. By giving `ProbeWeights()`, you are telling Primer3 how you want the resultant designs to be scored. But Primer3 complains because you didn't give it `ProbeParameters()` to score against -- that is what I meant by "optimal is undefined". 

Primer3 does not know what to do with the weights you provide it _in the absence of corresponding parameters_. Because each set of parameters is now defaulted to `None`, that is the context we are working within. 

Specifically, it is the `PRIMER_INTERNAL_WT_GC_PERCENT_*` weights that cause the problem, here is the error when you specify `PRIMER_INTERNAL_WT_GC_PERCENT_LT` and `PRIMER_INTERNAL_WT_GC_PERCENT_GT` but ask for left primers only:
```
PRIMER_ERROR=Hyb probe GC content is part of objective function while optimum gc_content is not defined
=
primer3_core: Hyb probe GC content is part of objective function while optimum gc_content is not defined
```

File: prymer/primer3/primer3.py (hunk 17 above)
Comment (emmcauley) [reply prymer/primer3/primer3.py:790]: no, none of this code is a new addition. I renamed one function, and refactored existing code into `_check_design_results` so I could use it for both primer and probe design

File: docs/overview.md [hunk 20]
```diff
@@ -3,8 +3,8 @@
 The `prymer` Python library is intended to be used for three main purposes:
 
 1. [Clustering targets](#clustering-targets) into larger amplicons prior to designing primers.
-2. [Designing primers](#designing-primers) (left or right) or primer pairs using Primer3 for each target from (1).
-3. [Build and Picking a set of primer pairs](#build-and-picking-primer-pairs) from the designed primer pairs produced in (2).
+2. [Designing](#designing-primers) primers (single and paired) and internal hybridization probes using Primer3 for each target from (1).
+3. [Build and Picking a set of primer pairs](#build-and-picking-primer-pairs) from the design candidates produced in (2).
```
Comment (emmcauley) [inline docs/overview.md:7]: we converged on `design()` instead of `design_oligos()` or `design_primers()`

File: prymer/api/oligo.py [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+
+    """
+
+    tm: float
+    penalty: float
+    span: Span
+    bases: Optional[str] = None
+    tail: Optional[str] = None
+
+    def __post_init__(self) -> None:
```
Comment (emmcauley) [inline prymer/api/oligo.py:107]: all of these methods derive from the original `Primer` class and are unchanged, just ported from `primer.py` to `oligo.py`

File: prymer/api/oligo_like.py [hunk 22]
```diff
@@ -124,7 +123,3 @@ def _strand_to_location_string(self) -> str:
             case _:  # pragma: no cover
                 # Not calculating coverage on this line as it should be impossible to reach
                 assert_never(f"Encountered unhandled Strand value: {self.span.strand}")
-
-
-PrimerLikeType = TypeVar("PrimerLikeType", bound=PrimerLike)
```
Comment (emmcauley) [inline prymer/api/oligo_like.py:129]: removed per PR comment

File: prymer/api/primer_pair.py [hunk 23]
```diff
@@ -37,7 +37,7 @@ class methods to represent a primer pair.  The primer pair is comprised of a lef
 Span(refname='chr1', start=21, end=100, strand=<Strand.POSITIVE: '+'>)
 
 >>> list(primer_pair)
-[Primer(name=None, tm=70.0, penalty=-123.0, span=Span(refname='chr1', start=1, end=20, strand=<Strand.POSITIVE: '+'>), bases='GGGGGGGGGGGGGGGGGGGG', tail=None), Primer(name=None, tm=70.0, penalty=-123.0, span=Span(refname='chr1', start=101, end=120, strand=<Strand.NEGATIVE: '-'>), bases='TTTTTTTTTTTTTTTTTTTT', tail=None)]
+[Oligo(name=None, tm=70.0, penalty=-123.0, span=Span(refname='chr1', start=1, end=20, strand=<Strand.POSITIVE: '+'>), bases='GGGGGGGGGGGGGGGGGGGG', tail=None), Oligo(name=None, tm=70.0, penalty=-123.0, span=Span(refname='chr1', start=101, end=120, strand=<Strand.NEGATIVE: '-'>), bases='TTTTTTTTTTTTTTTTTTTT', tail=None)]
```
Comment (emmcauley) [inline prymer/api/primer_pair.py:40]: because we stripped out the thermo-related attributes, I reverted the doc-test examples

File: prymer/primer3/primer3.py [hunk 24]
```diff
@@ -371,12 +365,25 @@ def design_primers(self, design_input: Primer3Input) -> Primer3Result:  # noqa:
                 f"Error, trying to use a subprocess that has already been "
                 f"terminated, return code {self._subprocess.returncode}"
             )
-
-        design_region: Span = self._create_design_region(
-            target_region=design_input.target,
-            max_amplicon_length=design_input.params.max_amplicon_length,
-            min_primer_length=design_input.params.min_primer_length,
-        )
+        design_region: Span
```
Comment (emmcauley) [inline prymer/primer3/primer3.py:368]: I kept this in, @tfenne flagging this for you because I wasn't sure if we decided to take it out.

File: prymer/primer3/primer3.py [hunk 25]
```diff
@@ -760,3 +744,44 @@ def _create_design_region(
...
+            primer_error = design_results["PRIMER_ERROR"]
+            raise ValueError(f"Primer3 returned an error: {primer_error}")
+        else:
+            raise ValueError(f"Primer3 did not return the count tag: {count_tag}")
+    count: int = int(maybe_count)
+
+    return count
+
+
+def _has_acceptable_dinuc_run(design_input: Primer3Input, oligo_design: Oligo) -> bool:
```
Comment (emmcauley) [inline prymer/primer3/primer3.py:764]: same -- existing code that I re-orged into a module-level function and renamed to be more descriptive

File: prymer/primer3/primer3_parameters.py (hunk 6 above)
Comment (emmcauley) [reply prymer/primer3/primer3_parameters.py]: these have been removed in the latest commit per our Slack discussion

File: prymer/primer3/primer3.py [hunk 26]
```diff
@@ -371,12 +365,25 @@ def design_primers(self, design_input: Primer3Input) -> Primer3Result:  # noqa:
...
+                    )
+                design_region = design_input.target
+            case DesignRightPrimersTask() | DesignLeftPrimersTask() | DesignPrimerPairsTask():
+                design_region = self._create_design_region(
+                    target_region=design_input.target,
+                    max_amplicon_length=design_input.primer_and_amplicon_params.max_amplicon_length,
+                    min_primer_length=design_input.primer_and_amplicon_params.min_primer_length,
+                )
+            case _ as unreachable:
+                assert_never(unreachable)
```
Comment (tfenne) [inline prymer/primer3/primer3.py:386]: Should this line get the same `pragma: unreachable` as the one below?

File: prymer/primer3/primer3_input.py (hunk 4 above)
Comment (tfenne) [reply prymer/primer3/primer3_input.py:115]: Circling back to this with one more thought.  Would it be hard to still have these defaulted on the input, but then depending on the task, only translate the appropriate weights to tags and send them to primer3?

**Review by tfenne — APPROVED** [review]
A couple of comments, but please squash and merge once you've addressed or told me why you won't address!
